src/highlevel/bidib_highlevel_util.c

```diff
@@ -224,3 +266,156 @@ void syslog_libbidib(int priority, const char *format, ...) {
+	uint32_t trackstate_reversers_mutex_lockedcount = 0;
+	uint32_t trackstate_trains_mutex_lockedcount = 0;
+	uint32_t trackstate_boosters_mutex_lockedcount = 0;
+	uint32_t trackstate_track_outputs_mutex_lockedcount = 0;
+	uint32_t bidib_state_boards_rwlock_lockedcount = 0;
+	
+	struct timespec start, end;
+	clock_gettime(CLOCK_MONOTONIC_RAW, &start);
+	
+	while (bidib_running) {
+		//sleep(2);
+		//struct timespec tv;
+		//clock_gettime(CLOCK_MONOTONIC, &tv);
+		/////TODO: Back to debug?
+		//syslog_libbidib(LOG_WARNING, "Heartbeat, time %ld.%.5ld", tv.tv_sec, tv.tv_nsec);
+		
+		// check the status of locked/unlocked mutexes.
+		iters++;
+		
+		if (pthread_rwlock_trywrlock(&bidib_state_trains_rwlock) == 0) {
+			pthread_rwlock_unlock(&bidib_state_trains_rwlock);
+		} else {
+			++bidib_state_trains_rwlock_lockedcount;
+		}
+		
+		if (pthread_mutex_trylock(&trackstate_accessories_mutex) == 0) {
+			pthread_mutex_unlock(&trackstate_accessories_mutex);
+		} else {
+			++trackstate_accessories_mutex_lockedcount;
+		}
+		
+		if (pthread_mutex_trylock(&trackstate_peripherals_mutex) == 0) {
+			pthread_mutex_unlock(&trackstate_peripherals_mutex);
+		} else {
+			++trackstate_peripherals_mutex_lockedcount;
+		}
+		
+		if (pthread_mutex_trylock(&trackstate_segments_mutex) == 0) {
+			pthread_mutex_unlock(&trackstate_segments_mutex);
+		} else {
+			++trackstate_segments_mutex_lockedcount;
+		}
+		
+		if (pthread_mutex_trylock(&trackstate_reversers_mutex) == 0) {
+			pthread_mutex_unlock(&trackstate_reversers_mutex);
+		} else {
+			++trackstate_reversers_mutex_lockedcount;
+		}
+		
+		if (pthread_mutex_trylock(&trackstate_trains_mutex) == 0) {
+			pthread_mutex_unlock(&trackstate_trains_mutex);
+		} else {
+			++trackstate_trains_mutex_lockedcount;
+		}
+		
+		if (pthread_mutex_trylock(&trackstate_boosters_mutex) == 0) {
+			pthread_mutex_unlock(&trackstate_boosters_mutex);
+		} else {
+			++trackstate_boosters_mutex_lockedcount;
+		}
+		
+		if (pthread_mutex_trylock(&trackstate_track_outputs_mutex) == 0) {
+			pthread_mutex_unlock(&trackstate_track_outputs_mutex);
+		} else {
+			++trackstate_track_outputs_mutex_lockedcount;
+		}
+		
+		if (pthread_rwlock_trywrlock(&bidib_state_boards_rwlock) == 0) {
+			pthread_rwlock_unlock(&bidib_state_boards_rwlock);
+		} else {
+			++bidib_state_boards_rwlock_lockedcount;
+		}
+		
+		usleep(5000); // 0.005s (1/100th)
+		//usleep(10000); // 0.01s (1/100th)
+		
+		// Report the mutex statistics every 2 seconds
+		if (iters == 400) {
+			clock_gettime(CLOCK_MONOTONIC_RAW, &end);
+			uint64_t elapsed_us = (end.tv_sec - start.tv_sec) * 1000000 + (end.tv_nsec - start.tv_nsec) / 1000;
+			
+			syslog_libbidib(LOG_WARNING, "Mutex statistics report for last 2 seconds (actual elapsed time: %llu, rounded to seconds: %llu):\n", 
+			       elapsed_us, elapsed_us / 1000000);
+			syslog_libbidib(LOG_WARNING, "bidib_state_trains_rwlock:   Wr-Locked %.2f percent of the time\n", 
```

Comment:
Yeah absolutely. I also fear that this is misleading due to the scheduler. Not sure how smart it is, but why would it schedule a thread that it might know will access a locked resource?

The one case where this should help, regarding the temporary lockup problem, is if that lockup would be caused by one thread holding a mutex for a long time. The lockups I've seen were often ~4s long or longer. If, for example, the lockup would be caused by one thread holding the board rwlock for 4 seconds, then this here should actually show a high percentage. But I agree, for other cases its misleading.   
Maybe I'll ask Michael Engel if he knows of a better way to profile lock/mutex usage.

Reply:
Doesn't look like you can attach a callback function every time the Pthread lock and unlock functions are called. The next best thing would be to create decorate the lock and unlock functions with performance counters.

Reply:
i.e., have a function that wraps the call to, e.g., `pthread_mutex_lock` and then do some measuring with that?    
Thinking about what exactly could be done in there...
- just logging timestamps and thread id -> the resulting log will not be very easy to overview.
- record the thread-id, lock and timestamp in some datastructure, then when the lock is released pop the record and save elsewhere some statistics on how long the thread held the lock? -> that's a lot of work
- ?

Reply:
Anyway, I'll remove the current sampling done in the heartbeat function, and mark this conversation as resolved. Actual profiling will have to be done another time.